### ISSUE TITLE
Allow Recursive setup of submodules with Terrafiles in them

### ISSUE THREAD

**Issue by anuj-modi** [devopsmakers/xterrafile#18]
`xterrafile` only initializes the top level of modules. For example, if my terraform uses module A and module A has a Terrafile which pulls in module B, `xterrafile` will correctly clone module A but not module B.

**Comment by pysysops** [comment]
Hi @anuj-modi I have been pondering this lately... Here's my (untested) thoughts and opinions: 

The idea of "vendoring" a module is that you `install` the required modules and commit them to Git so your module or infrastructure repo are an atomic unit with everything committed to it, much like this Git repo contains all of the go modules and dependencies vendored.

So, you produce some modules that make use of upstream modules for example pulling together an autoscaling module, a loadbalancer module and a security-group module to creat an opinionated "service" module. You use `xterrafile` to commit your upstream dependencies to your module repo. Of course, you have a CI process that runs some tests and release your module with `git tag ...`. You have a versioned, atomic, self contained module (that's why I strip the .git file out of modules).

In your infrastructure or product control repo, you add a `Terrafile` referencing your service module. `xterrafile install` will pull in your module version, you know it works, is tested and has all of it's dependencies contained in it.

Sure, you will probably end up with the same module duplicated in several of your service modules but it really doesn't take up that much space and the benefits I can see are that you can update your service modules independently regardless of what modules / versions you may be using in other modules or in your top level infrastructure repo.

I've worked with Terraform code that had a lot of upstream modules shared between service modules. If you wanted to update one of the shared modules then you'd have to make changes to all of your service modules which is unnecessarily wasteful and leads to big bang changes of Terraform code.

I don't think that iterating into modules and pulling their deps is wise, they should probably have everything they need in the module repo. I'm open to be persuaded otherwise or discuss this further though 😃

**Comment by lorengordon** [comment]
> So, you produce some modules that make use of upstream modules for example pulling together an autoscaling module, a loadbalancer module and a security-group module to creat an opinionated "service" module. You use xterrafile to commit your upstream dependencies to your module repo.

What if the upstream modules you use are also using modules? Any module someone uses could be using a module itself. It seems impractical to require them all to use `xterrafile` or otherwise vendor modules into their project... It makes more sense to me to use `xterrafile` only with a root module/config, and having `xterrafile` iterate through submodules to vendor them.

**Comment by pysysops** [comment]
I think I get you @lorengordon - but where would the modules be vendored to? the top level `vendor/modules`? `vendor/modules` within the upstream module? Perhaps the author of that module has decided to use a different location and a different source path in their module references?

I think if you write a module that requires upstream modules and you use a `Terrafile` / `xterrafile` approach to manage the upstream dependencies you'd probably want to keep your dependencies within the confines of your module. If other upstream modules don't use the `Terrafile` approach and have module sources written in code then `xterrafile` doesn't touch them, `terraform init` takes care of them.

Here's an example module that makes use of 3 upstream modules using the `Terrafile` approach. 
https://github.com/devopsmakers/tf-example-module it references the modules like:

```
module "this_asg" {
  source = "./vendor/modules/aws-asg"

  name = "${var.product}-${var.environment}-${var.service}"
  .....
```

The module contains everything it needs within it's repo, it can name the upstream modules whatever it wants and can be tested and developed on in isolation. Another module using a different version of an upstream module would not cause any issues.

A `terraform init` locates the modules in their appropriate place:
```
~/Projects/tf-example/dev/stg/services/app master !1       ⎈ docker-for-desktop
❯ terraform init
Initializing modules...
- app in ../../../../vendor/modules/asg-elb
- app.this_asg in ../../../../vendor/modules/asg-elb/vendor/modules/aws-asg
- app.this_elb in ../../../../vendor/modules/asg-elb/vendor/modules/aws-elb
- app.this_elb.elb in ../../../../vendor/modules/asg-elb/vendor/modules/aws-elb/modules/elb
- app.this_elb.elb_attachment in ../../../../vendor/modules/asg-elb/vendor/modules/aws-elb/modules/elb_attachment
- app.this_elb_sg in ../../../../vendor/modules/asg-elb/vendor/modules/aws-security-group
- app.this_instance_sg in ../../../../vendor/modules/asg-elb/vendor/modules/aws-security-group
```

and an example of how you might use it:

<img width="1049" alt="Screenshot 2019-10-15 at 22 15 31" src="https://user-images.githubusercontent.com/1253072/66870518-83190300-ef99-11e9-833d-94966fecfdfd.png">

I don't think there's any true consensus on nested Terraform modules (yet) so this is all interesting discussion.

I wouldn't expect to see module sources within modules breaking out of the module.

If you're developing your own modules and don't want to vendor modules / commit them to git you could always do something like this:

```
CURRENT_DIR=$(pwd)
for tf in $(find . -iname Terrafile); do
  cd $(dirname ${CURRENT_DIR}/${tf})
  xterrafile install
done
cd ${CURRENT_DIR}
```

**Comment by lorengordon** [comment]
Yeah, trying to avoid the need to use `terraform init` to fetch the remote, nested module. But you're right, the scenario I'm thinking of would require re-writing the `source` argument for the upstream modules, or perhaps making judicious use of [overrides](https://www.terraform.io/docs/configuration/override.html).

**Comment by pysysops** [comment]
I have worked with infrastructures where upstream modules were all shared in numerous other local modules and service modules and it was painful. Updating a single module could mean having to change many, many different modules.

I've also worked somewhere that used git submodules and had a lot of cross referencing modules. That was 💩

**Comment by pysysops** [comment]
`terraform init` is pretty much a requirement before a `plan` and `apply`, it does all the backend initialisation and locking, installs plugins and providers... https://www.terraform.io/docs/commands/init.html

What the `Terrafile` approach is aimed at helping with is making all of your modules clearly visible in a single file instead of in `source` arguments throughout the Terraform code base which often grows and becomes painful to manage all of the instances of the `source` argument.

I'm still in the early stages of forming opinion around Terraform modules and managing them in a simple way... I started a post here: https://www.pysysops.com/2019/10/03/Taming-Terraform-with-Modules.html and was going to post about nested modules this week.

Anton Babenko (@antonbabenko) has a lot of ideas and experience when it comes to Terraform modules.

**Comment by lorengordon** [comment]
Sorry, I'm not trying to be obtuse. `terraform init` is definitely required for the backend. It's also fine for retrieving providers (can also use `terraform-bundle` to pre-stage the providers). I'm not trying to avoid `terraform init` _entirely_. It's specifically just the part of `terraform init` that retrieves modules from remote sources that I'd like to avoid, and what has me interested in `xterrafile`.

**Comment by pysysops** [comment]
To answer the question or suggestion definitively:

Until I've done more in terms of reasoning around vendored modules / nested module dependencies I won't be considering a recursive feature. Especially as it can be easily achieved with a simple for loop on a find command at this point in time.

It _might_ be re-considered in the future, just not in the near future.
